### ISSUE TITLE
statuscmd with signals

### ISSUE THREAD

**Issue by dazWiLLiE** [bakkeby/dwm-flexipatch#114]
I can't get it to work, when I click on the bar, dwmblock exits with:
`[1]    702746 user-defined signal 1  dwmblocks`

I've fought with this for a few days now, with minimal patches, and even tried with flexipatch 1.0 thinking it perhaps was the new bar problem. Tried dwmblocks with only statuscmd patch, different versions of dwmblocks (including Luke's).. same result.

With Luke's dwm + dwmblocks it works fine.

Patchlist:
```
#define BAR_LTSYMBOL_PATCH 1
#define BAR_STATUS_PATCH 1
#define BAR_TAGS_PATCH 1
#define BAR_WINTITLE_PATCH 1
#define BAR_TITLE_LEFT_PAD_PATCH 1

#define TILE_LAYOUT 1
#define MONOCLE_LAYOUT 1

#define BAR_DWMBLOCKS_PATCH 1
#define BAR_STATUSCMD_PATCH 1
#define RESTARTSIG_PATCH 1
#define XRDB_PATCH 1

#define BAR_STATUS2D_PATCH 1
#define BAR_SYSTRAY_PATCH 1
#define BAR_ALPHA_PATCH 1
#define BAR_COLOR_EMOJI_PATCH 1
#define ATTACHBOTTOM_PATCH 1
#define AUTOSTART_PATCH 1
#define NO_TRANSPARENT_BORDERS_PATCH 1
#define SCRATCHPADS_PATCH 1
#define SHIFTVIEW_PATCH 1
#define STACKER_PATCH 1
#define STICKY_PATCH 1
#define SWALLOW_PATCH 1
#define VANITYGAPS_PATCH 1
#define BSTACK_LAYOUT 1
#define CENTEREDMASTER_LAYOUT 1
#define CENTEREDFLOATINGMASTER_LAYOUT 1
#define DECK_LAYOUT 1
#define FIBONACCI_DWINDLE_LAYOUT 1
#define FIBONACCI_SPIRAL_LAYOUT 1
```

config.def.h

```
static const BarRule barrules[] = {
        /* monitor  bar    alignment         widthfunc                drawfunc                clickfunc                name */
        { 'A',      0,     BAR_ALIGN_RIGHT,  width_status2d,          draw_status2d,          click_statuscmd,         "status2d" },

static Button buttons[] = {
        /* click                event mask           button          function        argument */
        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 1} },
        { ClkStatusText,        0,                   Button2,        sigdwmblocks,   {.i = 3} },

```

blocks.def.h

```
static const Block blocks[] = {
        /*Icon*/        /*Command*/             /*Update Interval*/     /*Update Signal*/
        {"",    "test1.sh",     5,     1},
        {"",    "test2.sh",    60,     3},
};
```

Edit: Oops, I meant "dwmblocks", not "dwmbar". I was tired, I guess.

**Comment by dazWiLLiE** [comment]
Nevermind, I built dwmblocks again with the [patch](https://gist.github.com/IGeraGera/e4a5583b91b3eec2e81fdceb44dea717) linked directly on the front [dwmblocks github page](https://github.com/torrinfail/dwmblocks) instead of the patch on the [suckless page](https://dwm.suckless.org/patches/statuscmd/dwm-statuscmd-status2d-20210405-60bb3df.diff).

Works fine now.

**Comment by bakkeby** [comment]
I'm not entirely sure about this, the original author replaced the diff on the suckless page recently.

https://git.suckless.org/sites/commit/6d161b55f4cd0079d0d2353728979e9762f24ee1.html
https://git.suckless.org/sites/commit/1f20d8cefb96cdf64300ddb9168a2e2b6c273743.html

**Comment by dazWiLLiE** [comment]
Ok, so dwm-flexipatch with dwmblocks patched with [this patch](https://git.suckless.org/sites/commit/1f20d8cefb96cdf64300ddb9168a2e2b6c273743.html) should work?

I'm quite sure that thats one of the patches I tried.

**Comment by bakkeby** [comment]
What exactly is it that is not working for you?

The only main difference between the patches for dwmblocks is that in newer versions they seem to set the environment variable `BUTTON` to indicate what button was pressed, whereas older versions used the `BLOCK_BUTTON` environment variable instead.

This is isolated to dwmblocks though and shouldn't make a difference for dwm, unless they also changed the kill signal mechanism.

Which environment variable are you checking in your script?

**Comment by dazWiLLiE** [comment]
I cloned dwmblocks again, applying the [this patch](https://dwm.suckless.org/patches/statuscmd/dwmblocks-statuscmd-20210402-96cbb45.diff).

When I click on any of the blocks, dwmblocks exits with:
`[1]    669673 user-defined signal 1  dwmblocks`

With dwmblocks patched with [this patch](https://gist.github.com/IGeraGera/e4a5583b91b3eec2e81fdceb44dea717), it works.

**Comment by dazWiLLiE** [comment]
Example script:
```
#!/bin/sh

case $BUTTON in
        1) notify-send "test 1" ;;
        2) notify-send "test 2" ;;
        3) notify-send "test 3" ;;
esac

echo "text"

```

config.def.h:

```
/* button definitions */
/* click can be ClkTagBar, ClkLtSymbol, ClkStatusText, ClkWinTitle, ClkClientWin, or ClkRootWin */
static Button buttons[] = {
        /* click                event mask           button          function        argument */
        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 1} },        //dwmcpu
//        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 2} },        //dwmmemory
        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 3} },        //dwmnet
//        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 4} },        //dwmtemp
        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 5} },        //dwmdisk
        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 6} },        //dwmupdates
        { ClkStatusText,        0,                   Button2,        sigdwmblocks,   {.i = 6} },        //dwmupdates
        { ClkStatusText,        0,                   Button3,        sigdwmblocks,   {.i = 6} },        //dwmupdates
        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 7} },        //dwmvol
        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 8} },        //dwmweather
        { ClkStatusText,        0,                   Button2,        sigdwmblocks,   {.i = 8} },        //dwmweather
        { ClkStatusText,        0,                   Button1,        sigdwmblocks,   {.i = 9} },        //dwmclock
        { ClkLtSymbol,          0,                   Button1,        setlayout,      {0} },
        { ClkLtSymbol,          0,                   Button3,        setlayout,      {.v = &layouts[2]} },
        { ClkWinTitle,          0,                   Button2,        zoom,           {0} },
        { ClkClientWin,         MODKEY,              Button1,        movemouse,      {0} },
        { ClkClientWin,         MODKEY,              Button2,        togglefloating, {0} },
        { ClkClientWin,         MODKEY,              Button3,        resizemouse,    {0} },
        { ClkTagBar,            0,                   Button1,        view,           {0} },
        { ClkTagBar,            0,                   Button3,        toggleview,     {0} },
        { ClkTagBar,            MODKEY,              Button1,        tag,            {0} },
        { ClkTagBar,            MODKEY,              Button3,        toggletag,      {0} },
};
```

blocks.def.h

```
//Modify this file to change what commands output to your statusbar, and recompile using the make command.
static const Block blocks[] = {
        /*Icon*/        /*Command*/             /*Update Interval*/     /*Update Signal*/
        {"",    "dwmcpu",          3,      1},
//      {"",    "dwmmemory",      60,      2},
        {"",    "dwmnet",          1,      3},
//      {"",    "dwmtemp",        60,      4},
        {"",    "dwmdisk",       120,      5},
        {"",    "dwmupdates",      0,      6},
        {"",    "dwmvol",          0,      7},
        {"",    "dwmweather",   3600,      8},
        {"",    "dwmclock",       60,      9},
};

//sets delimeter between status commands. NULL character ('\0') means no delimeter.
static char delim[] = " | ";
static unsigned int delimLen = 5;
```

**Comment by bakkeby** [comment]
I have had a closer look and the author has indeed changed the signalling mechanism from using the user defined signal 1 (SIGUSR1) to using specific real-time signals (SIGRTMIN) instead.

In practice the two are not compatible with each other which means you have to have the same changes in dwm as you have in dwmblocks. I'll look into adding support for both.

**Comment by bakkeby** [comment]
I upgraded the statuscmd patch (and the now named statuscmd-nosignal patch). I added a compatibility option which is `BAR_DWMBLOCKS_SIGUSR1_PATCH` to fall back to using SIGUSR1 signals rather than SIGRTMIN (for those with a dwmblocks build that predates the new patch).

Note that the configuration has also changed in this respect, replacing `sigdwmblocks` calls with `sigstatusbar` and the STATUSBAR macro defining the name of the dwmblocks executable.

**Comment by dazWiLLiE** [comment]
Great, thanks :)

I had some time to test it out now.
However, I don't know if its me, dwmblocks or your new patch.
It works, but it seems that I get a lot of "noise".
I tested with:

```
case $BUTTON in
   *) notify-send $BUTTON
esac
```

left clicked, and got `1, 9 ,7 ,3 ,6 ,8` as result.
If i right click I get 6 as a result.

In dwm/config.def.h I have only used Button1 for this script:
`{ ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 7} },        //dwmvol`

**Comment by bakkeby** [comment]
Oh, ok, I think it is just a misunderstanding regarding the value you pass in here. `{.i = 7}` is an indicator just to say that you clicked on button 7, it doesn't have anything to do with the status or which block it is. Try leaving these as-is so that Button1 passes `{.i = 1}`, Button3 (right click) passes `{.i = 3}`, etc. and I think it should be fine.

**Comment by dazWiLLiE** [comment]
Hmm. So how does dwmblocks know which block to update? I guess that "Update Signal" (dwmblocks/blocks.def.h) only applies to manually signaling with SIGRTMIN then?

I tried this now. So, dwm/config.h:
```
static Button buttons[] = {
        /* click                event mask           button          function        argument */
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 1 dwmcpu
//      { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 2 dwmmemory
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 3 dwmnet
//      { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 4 dwmtemp
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 5 dwmdisk
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 6 dwmupdates
        { ClkStatusText,        0,                   Button2,        sigstatusbar,   {.i = 2} },        // 6 dwmupdates
        { ClkStatusText,        0,                   Button3,        sigstatusbar,   {.i = 3} },        // 6 dwmupdates
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 7 dwmvol
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 8 dwmweather
        { ClkStatusText,        0,                   Button2,        sigstatusbar,   {.i = 2} },        // 8 dwmweather
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 9 dwmclock
```

dwmblocks/blocks.def.h:
```
//Modify this file to change what commands output to your statusbar, and recompile using the make command.
static const Block blocks[] = {
        /*Icon*/        /*Command*/             /*Update Interval*/     /*Update Signal*/
//        {"",    "/home/magnus/.scripts/dwmscripts/dwmcpu",          1,      1},
        {"",    "/home/magnus/.scripts/dwmscripts/dwmcpu",          3,      1},
//      {"",    "/home/magnus/.scripts/dwmscripts/dwmmemory",      60,      2},
        {"",    "/home/magnus/.scripts/dwmscripts/dwmnet",          1,      3},
//      {"",    "/home/magnus/.scripts/dwmscripts/dwmtemp",        60,      4},
        {"",    "/home/magnus/.scripts/dwmscripts/dwmdisk",       120,      5},
        {"",    "/home/magnus/.scripts/dwmscripts/dwmupdates",      0,      6},
        {"",    "/home/magnus/.scripts/dwmscripts/dwmvol",          0,      7},
        {"",    "/home/magnus/.scripts/dwmscripts/dwmweather",   3600,      8},
        {"",    "/home/magnus/.scripts/dwmscripts/dwmclock",       60,      9},
};

//sets delimeter between status commands. NULL character ('\0') means no delimeter.
static char delim[] = " | ";
static unsigned int delimLen = 5;
```
/dwmupdates:
```
#!/bin/sh

case $BUTTON in
        1) st -e sb-popupgrade ;;
        3) notify-send "$(/usr/bin/pacman -Qu)" ;;
        *) notify-send $BUTTON ;;
esac

pacman -Qu | grep -Fcv "[ignored]" | sed "s/^/📦 /;s/^📦0$//g"

```

If I left click on the "dwmupdates" block, st spawns 7 times. Right click behaves correctly.

Am I missing something else? :zany_face:

**Comment by bakkeby** [comment]
There are two stages for handling button clicks:
   - dwm will work out what you clicked on based on the window and x and y position of the mouse click and
   - dwm will use that information to trigger function bindings (if any)

The bindings you refer to is for the later part.

So you have set up a series of button handlers right.

```c
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 1 dwmcpu
//      { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 2 dwmmemory
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 3 dwmnet
//      { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 4 dwmtemp
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 5 dwmdisk
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 6 dwmupdates
        { ClkStatusText,        0,                   Button2,        sigstatusbar,   {.i = 2} },        // 6 dwmupdates
        { ClkStatusText,        0,                   Button3,        sigstatusbar,   {.i = 3} },        // 6 dwmupdates
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 7 dwmvol
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 8 dwmweather
        { ClkStatusText,        0,                   Button2,        sigstatusbar,   {.i = 2} },        // 8 dwmweather
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 9 dwmclock
```

You click on the bar, and it is determined that the part of the bar you clicked on was the status text so we have a match for `ClkStatusText`.

You clicked using the left mouse button, so we have a match for `Button1`.

The code that handles button clicks loops through all of these button handlers and will trigger all of these seven bindings:
```c
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 1 dwmcpu
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 3 dwmnet
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 5 dwmdisk
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 6 dwmupdates
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 7 dwmvol
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 8 dwmweather
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // 9 dwmclock
```

Hence st spawns seven times in your case.

What you should have is this, as per the default configuration:

```c
        { ClkStatusText,        0,                   Button1,        sigstatusbar,   {.i = 1} },        // left click
        { ClkStatusText,        0,                   Button2,        sigstatusbar,   {.i = 2} },        // middle click
        { ClkStatusText,        0,                   Button3,        sigstatusbar,   {.i = 3} },        // right click
        { ClkStatusText,        0,                   Button4,        sigstatusbar,   {.i = 4} },        // scroll wheel up
        { ClkStatusText,        0,                   Button5,        sigstatusbar,   {.i = 5} },        // scroll wheel down
```

In your configuration for example you have set that the dwmnet script has the update signal of 3, so when dwmblocks combines all of the output form your scripts it will prefix the dwmnet output with a 0x03 character (ETX / end of text according to the ASCII table). The examples in https://dwm.suckless.org/patches/statuscmd/ are quite good.

When you initially click on the bar it will calculate first what part of the bar you clicked on. When you click on the status it will work out what block you clicked on based on these non-printable characters (characters less than the space character (30)). So for the dwmnet block it will spot the 0x03 marker and it will store that in an internal variable.

When the button binding fires the `sigstatusbar` function is called. This function will send a kill signal SIGRTMIN+3 with the value of 1 to dwmblocks.

dwmblocks in turn reacts by setting the `BUTTON` environment variable to the value of 1, and since it finds that it has a module with an update signal of 3 it will re-run the dwmnet script for the output.

The dwmnet script spots that the BUTTON environment variable is set and acts accordingly depending on what button was pressed, 1, 2, 3, 4, 5, etc.

In other words if you have set up bindings for buttons 1 through 5 then that applies for all of your scripts. It is up to the scripts whether they handle those kind of button clicks or not.

I tend to use Button8 (back button) to open the script itself in my editor of choice.

TLDR; you don't need to mess with the button bindings unless you need to add more buttons.

**Comment by dazWiLLiE** [comment]
Wow, I completely misunderstood the argument to the sigstatusbar function. Thanks for the patch and explanation, everything works as expected now :)
